File: lib/core/utils/link_utils.dart
import 'package:url_launcher/url_launcher.dart';

class LinkUtils {
  LinkUtils._();

  static Future<void> triggerWhatsApp(String? phone, [String? recruiterName]) async {
    if (phone == null || phone.isEmpty) return;
    
    // Clean phone number
    String cleanPhone = phone.replaceAll(RegExp(r'[^0-9]'), '');
    if (cleanPhone.length == 10) {
      cleanPhone = '91$cleanPhone';
    }
    
    final text = recruiterName != null && recruiterName.isNotEmpty
        ? 'Hi ${recruiterName.trim()}, '
        : '';
    final encodedText = Uri.encodeComponent(text);
    
    final uri = Uri.parse('whatsapp://send?phone=$cleanPhone&text=$encodedText');
    final fallbackUri = Uri.parse('https://api.whatsapp.com/send?phone=$cleanPhone&text=$encodedText');

    if (await canLaunchUrl(uri)) {
      await launchUrl(uri, mode: LaunchMode.externalApplication);
    } else {
      await launchUrl(fallbackUri, mode: LaunchMode.externalApplication);
    }
  }

  static Future<void> triggerSMS(String? phone) async {
    if (phone == null || phone.isEmpty) return;
    final cleanPhone = phone.replaceAll(RegExp(r'[^0-9+]'), '');
    final uri = Uri.parse('sms:$cleanPhone');
    final fallbackUri = Uri.parse('https://messages.google.com/web/conversations/new?phone=$cleanPhone');

    if (await canLaunchUrl(uri)) {
      await launchUrl(uri, mode: LaunchMode.externalApplication);
    } else {
      await launchUrl(fallbackUri, mode: LaunchMode.externalApplication);
    }
  }

  static Future<void> triggerGmail(String? email, [String? recruiterName]) async {
    if (email == null || email.isEmpty) return;
    final subject = Uri.encodeComponent("Job Application Status");
    final body = recruiterName != null && recruiterName.isNotEmpty
        ? Uri.encodeComponent("Hi ${recruiterName.trim()},\n\n")
        : '';
        
    final uri = Uri.parse('mailto:$email?subject=$subject&body=$body');
    await launchUrl(uri, mode: LaunchMode.externalApplication);
  }

  static Future<void> triggerPhone(String? phone) async {
    if (phone == null || phone.isEmpty) return;
    final cleanPhone = phone.replaceAll(RegExp(r'[^0-9+]'), '');
    final uri = Uri.parse('tel:$cleanPhone');
    await launchUrl(uri, mode: LaunchMode.externalApplication);
  }

  static Future<void> launchWeb(String? url) async {
    if (url == null || url.isEmpty) return;
    var cleanUrl = url.trim();
    if (!cleanUrl.startsWith('http://') && !cleanUrl.startsWith('https://')) {
      cleanUrl = 'https://$cleanUrl';
    }
    final uri = Uri.parse(cleanUrl);
    await launchUrl(uri, mode: LaunchMode.externalApplication);
  }
}

File: src/lib/linkUtils.ts
/**
 * Utility functions to generate and trigger native deep links for calling, WhatsApp, SMS, and Gmail on mobile devices.
 */

export function isMobileDevice(): boolean {
  if (typeof window === 'undefined') return false;
  const userAgent = window.navigator.userAgent;
  const isTouchDevice = 'ontouchstart' in window || window.navigator.maxTouchPoints > 0;
  return /Mobi|Android|iPhone|iPad|iPod/i.test(userAgent) || (isTouchDevice && !/Windows/i.test(userAgent));
}

export function triggerWhatsApp(phone: string | undefined, recruiterName?: string): void {
  if (typeof window === 'undefined' || !phone) return;
  let cleanPhone = phone.replace(/[^0-9]/g, '');
  if (cleanPhone.length === 10) {
    cleanPhone = '91' + cleanPhone;
  }
  const text = recruiterName ? `Hi ${recruiterName.trim()}, ` : '';
  const encodedText = text ? `&text=${encodeURIComponent(text)}` : '';

  if (isMobileDevice()) {
    // Mobile native app deep link
    window.location.href = `whatsapp://send?phone=${cleanPhone}${encodedText}`;
  } else {
    // Desktop universal web link fallback
    window.open(`https://api.whatsapp.com/send?phone=${cleanPhone}${encodedText}`, '_blank', 'noopener,noreferrer');
  }
}

export function triggerSMS(phone: string | undefined, _recruiterName?: string): void {
  if (typeof window === 'undefined' || !phone) return;
  void _recruiterName;
  const cleanPhone = phone.replace(/[^0-9+]/g, '');

  if (isMobileDevice()) {
    window.location.href = `sms:${cleanPhone}`;
  } else {
    window.open(`https://messages.google.com/web/conversations/new?phone=${cleanPhone}`, '_blank', 'noopener,noreferrer');
  }
}

export function triggerGmail(email: string | undefined, recruiterName?: string): void {
  if (typeof window === 'undefined' || !email) return;
  const subject = encodeURIComponent("Job Application Status");
  const body = recruiterName ? encodeURIComponent(`Hi ${recruiterName.trim()},\n\n`) : '';
  window.location.href = `mailto:${email}?subject=${subject}${body ? `&body=${body}` : ''}`;
}

export function triggerPhone(phone: string | undefined): void {
  if (typeof window === 'undefined' || !phone) return;
  const cleanPhone = phone.replace(/[^0-9+]/g, '');
  window.location.href = `tel:${cleanPhone}`;
}

// Deprecated: Retained for backward compatibility during build phase
export function getWhatsAppLink(phone: string, recruiterName?: string): string {
  let cleanPhone = phone.replace(/[^0-9]/g, '');
  if (cleanPhone.length === 10) cleanPhone = '91' + cleanPhone;
  const text = recruiterName ? `Hi ${recruiterName.trim()}, ` : '';
  const encodedText = text ? `&text=${encodeURIComponent(text)}` : '';
  if (isMobileDevice()) return `whatsapp://send?phone=${cleanPhone}${encodedText}`;
  return `https://api.whatsapp.com/send?phone=${cleanPhone}${encodedText}`;
}

export function getGmailLink(email: string, recruiterName?: string): string {
  const subject = encodeURIComponent("Job Application Status");
  const body = recruiterName ? encodeURIComponent(`Hi ${recruiterName.trim()},\n\n`) : '';
  return `mailto:${email}?subject=${subject}${body ? `&body=${body}` : ''}`;
}

export function getPhoneLink(phone: string): string {
  const cleanPhone = phone.replace(/[^0-9+]/g, '');
  return `tel:${cleanPhone}`;
}

export function getGoogleMessagesLink(phone: string, _recruiterName?: string): string {
  void _recruiterName;
  const cleanPhone = phone.replace(/[^0-9+]/g, '');
  if (isMobileDevice()) {
    return `sms:${cleanPhone}`;
  }
  return `https://messages.google.com/web/conversations/new?phone=${cleanPhone}`;
}
